From 015cf2049c6cbd90ebdd219e391735459b9a4c8d Mon Sep 17 00:00:00 2001
From: Kaarle Ritvanen <kaarle.ritvanen@datakunkku.fi>
Date: Wed, 1 Jan 2014 03:51:56 +0200
Subject: [PATCH] core/socket: close TCP server fd when spawning

If server fd is not closed, it is not possible to restart uwsgi if the
spawned process is still running as it keeps the address/port bound.
---
 core/socket.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/core/socket.c b/core/socket.c
index 0696eff..2325cbf 100644
--- a/core/socket.c
+++ b/core/socket.c
@@ -674,6 +674,8 @@ int bind_to_tcp(char *socket_name, int listen_queue, char *tcp_port) {
 		return -1;
 	}
 	
+	if (fcntl(serverfd, F_SETFD, FD_CLOEXEC) < 0) uwsgi_error("fcntl()");
+
 	if (uwsgi.so_sndbuf) {
 		socklen_t sndbuf = (socklen_t) uwsgi.so_sndbuf;
 		if (setsockopt(serverfd, SOL_SOCKET, SO_SNDBUF, &sndbuf, sizeof(socklen_t)) < 0) {
-- 
1.8.4.2

